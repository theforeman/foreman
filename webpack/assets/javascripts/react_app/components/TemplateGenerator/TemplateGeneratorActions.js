/* eslint-disable promise/prefer-await-to-then */
import { saveAs } from 'file-saver';
import { API } from '../../redux/API';

import {
  TEMPLATE_GENERATE_REQUEST,
  TEMPLATE_GENERATE_POLLING,
  TEMPLATE_GENERATE_SUCCESS,
  TEMPLATE_GENERATE_FAILURE,
} from './TemplateGeneratorConstants';

const pollingInterval = 3000;

export const generateTemplate = (url, templateInputData) => dispatch => {
  dispatch({
    type: TEMPLATE_GENERATE_REQUEST,
    payload: { ...templateInputData },
  });
  return API.post(url, templateInputData)
    .then(({ data }) => {
      dispatch(pollReportData(data.data_url));
    })
    .catch(error =>
      dispatch({
        type: TEMPLATE_GENERATE_FAILURE,
        payload: { error, item: templateInputData },
      })
    );
};

const _downloadFile = response => {
  const blob = new Blob([response.data], {
    type: response.headers['content-type'],
  });
  const filename = response.headers['content-disposition'].match(
    /filename="(.*)"/
  );
  saveAs(blob, (filename && filename[1]) || 'report.txt');
};

const _getErrors = errorResponse => {
  if (!errorResponse || !errorResponse.data) return null;
  if (errorResponse.status === 422) return errorResponse.data.errors;
  if (errorResponse.data.error) return [errorResponse.data.error]; // most of >500
  return [errorResponse.data];
};

export const pollReportData = pollUrl => dispatch => {
  dispatch({ type: TEMPLATE_GENERATE_POLLING, payload: { url: pollUrl } });

  return API.get(pollUrl, { responseType: 'blob' })
    .then(response => {
      if (response.status === 200) {
        dispatch({ type: TEMPLATE_GENERATE_SUCCESS, payload: {} });
        _downloadFile(response);
      } else if (pollingInterval) {
        setTimeout(() => dispatch(pollReportData(pollUrl)), pollingInterval);
      }
    })
    .catch(error => {
      dispatch({
        type: TEMPLATE_GENERATE_FAILURE,
        payload: { error, messages: _getErrors(error.response) },
      });
    });
};
