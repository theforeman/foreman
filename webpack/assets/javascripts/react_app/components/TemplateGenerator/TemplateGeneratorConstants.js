export const TEMPLATE_GENERATE_REQUEST = 'TEMPLATE_GENERATE_REQUEST';
export const TEMPLATE_GENERATE_POLLING = 'TEMPLATE_GENERATE_POLLING';
export const TEMPLATE_GENERATE_SUCCESS = 'TEMPLATE_GENERATE_SUCCESS';
export const TEMPLATE_GENERATE_FAILURE = 'TEMPLATE_GENERATE_FAILURE';
