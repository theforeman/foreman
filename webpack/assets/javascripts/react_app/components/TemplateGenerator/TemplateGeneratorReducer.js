import Immutable from 'seamless-immutable';

import {
  TEMPLATE_GENERATE_REQUEST,
  TEMPLATE_GENERATE_POLLING,
  TEMPLATE_GENERATE_SUCCESS,
  TEMPLATE_GENERATE_FAILURE,
} from './TemplateGeneratorConstants';

const initialState = Immutable({
  scheduleInProgress: false,
  polling: false,
  dataUrl: null,
});

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case TEMPLATE_GENERATE_REQUEST:
      return state.set('scheduleInProgress', true);
    case TEMPLATE_GENERATE_POLLING:
      return state.merge({
        scheduleInProgress: false,
        dataUrl: payload.url,
        polling: true,
      });
    case TEMPLATE_GENERATE_FAILURE:
      return state.merge({
        scheduleInProgress: false,
        polling: false,
        generatingError: payload.error.message,
        generatingErrorMessages: payload.messages,
      });
    case TEMPLATE_GENERATE_SUCCESS:
      return state.merge({
        scheduleInProgress: false,
        polling: false,
      });
    default:
      return state;
  }
};
