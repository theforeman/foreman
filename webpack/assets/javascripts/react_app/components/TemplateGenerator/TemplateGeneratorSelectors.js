const selectTemplates = state => state.templates;

const selectGeneratingPropsFromTemplates = ({
  polling,
  dataUrl,
  generatingError,
  generatingErrorMessages,
}) => ({
  polling,
  dataUrl,
  generatingError,
  generatingErrorMessages,
});

export const selectGeneratingProps = state =>
  selectGeneratingPropsFromTemplates(selectTemplates(state));
