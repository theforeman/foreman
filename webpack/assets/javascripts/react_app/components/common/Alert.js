import React from 'react';
import Icon from './Icon';
import { ALERT_CSS } from '../../constants';

const Alert = (props) => {
  let classNames = ALERT_CSS[props.type];

  if (props.css) {
    classNames += ' ' + props.css;
  }

  if (props.dismissable) {
    classNames += ' ' + ALERT_CSS.dismissable;
  }

  const closeBtn = (
    <button className="close" data-dismiss="alert" aria-hidden="true">
      <Icon type="close" />
    </button>
  );

  return (
    <div className={classNames}>
      {props.dismissable && closeBtn}
      {props.children}
    </div>
  );
};

Alert.defaultProps = { dismissable: false, css: null, type: 'info' };
export default Alert;
