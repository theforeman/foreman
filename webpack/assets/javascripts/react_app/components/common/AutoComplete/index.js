import React, { Component } from 'react';
import Downshift from 'downshift';
import { Dropdown, MenuItem, InputGroup, FormControl, Button } from 'patternfly-react';
import PropTypes from 'prop-types';

export const renderItems = ({
  items, highlightedIndex, selectedItem, getItemProps,
}) => (
  <Dropdown.Menu
    style={{
      display: 'block',
      position: 'absolute',
      left: 'inherit',
      marginTop: 0,
      top: 'auto',
    }}
  >
    {items.map((item, index) => (
      <MenuItem
        {...getItemProps({
          index,
          item,
          active: highlightedIndex === index,
          onClick: (e) => {
            // At this point the event.defaultPrevented
            // is already set to true by react-bootstrap
            // MenuItem. We need to set it back to false
            // So downshift will execute it's own handler
            e.defaultPrevented = false;
          },
        })}
        key={item}
      >
        {item}
      </MenuItem>
    ))}
  </Dropdown.Menu>
);

class AutoComplete extends Component {
  constructor(props) {
    super(props);

    this.state = {
      inputValue: '',
    };
  }

  render() {
    const {
      onSearch, labelText, onInputUpdate, items, actionText, ...rest
    } = this.props;

    return (
      <Downshift
        onStateChange={({ inputValue }) => {
          if (typeof inputValue === 'string') {
            onInputUpdate(inputValue);
            this.setState({ inputValue });
          }
        }}
        defaultHighlightedIndex={0}
        selectedItem={this.state.inputValue}
        {...rest}
      >
        {({
          getInputProps,
          getItemProps,
          getLabelProps,
          isOpen,
          inputValue,
          highlightedIndex,
          getRootProps,
          selectedItem,
        }) => (
          <div>
            {labelText && <label {...getLabelProps()}>{labelText}</label>}
            <InputGroup>
              <FormControl type="text" {...getInputProps()} />
              <InputGroup.Button>
                <Button onClick={() => onSearch(inputValue)}>{actionText}</Button>
              </InputGroup.Button>
            </InputGroup>

            {isOpen && items.length
              ? renderItems({
                  items,
                  highlightedIndex,
                  selectedItem,
                  getItemProps,
                })
              : null}
          </div>
        )}
      </Downshift>
    );
  }
}

AutoComplete.propTypes = {
  items: PropTypes.arrayOf(PropTypes.string).isRequired,
  onInputUpdate: PropTypes.func.isRequired,
  onSearch: PropTypes.func.isRequired,
  labelText: PropTypes.string,
  actionText: PropTypes.string,
};

AutoComplete.defaultProps = {
  labelText: '',
  actionText: 'Search',
};

export default AutoComplete;
