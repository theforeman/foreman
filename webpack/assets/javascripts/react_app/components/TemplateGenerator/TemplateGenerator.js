import React from 'react';
import PropTypes from 'prop-types';
import { Alert, Button } from 'patternfly-react';

import { sprintf, translate as __ } from '../../../react_app/common/I18n';
import AlertBody from '../common/Alert/AlertBody';

const pollingMsg = `
  Report %s is now being generated, the download will start once it's done.
  You can come to this page later to get the results. The result is available for 24 hours.
`;
const doneMsg = `
  Generating of the report %s has been completed.
  Download should start automatically.
  In case it does not, please use the download button below.
`;

const getAlert = (type, msg) => (
  <Alert type={type} title={__('Generating a report')}>
    <AlertBody message={msg} />
  </Alert>
);

class TemplateGenerator extends React.Component {
  getError() {
    const { generatingError, generatingErrorMessages } = this.props;
    const errors =
      generatingErrorMessages &&
      generatingErrorMessages.map(e => e.message).join('\n');

    return errors || generatingError;
  }

  renderAlert() {
    const {
      polling,
      data: { templateName },
    } = this.props;
    const error = this.getError();
    if (polling) return getAlert('info', sprintf(pollingMsg, templateName));
    if (error) return getAlert('error', error);
    return getAlert('success', sprintf(doneMsg, templateName));
  }

  render() {
    const { polling, dataUrl, generatingError } = this.props;

    if (!dataUrl && !polling) return null;

    return (
      <React.Fragment>
        {this.renderAlert()}
        {!polling && !generatingError && (
          <Button bsStyle="primary" href={dataUrl} target="_blank">
            {__('Download')}
          </Button>
        )}
      </React.Fragment>
    );
  }
}

TemplateGenerator.propTypes = {
  data: PropTypes.shape({
    templateName: PropTypes.string.isRequired,
  }).isRequired,
  polling: PropTypes.bool,
  dataUrl: PropTypes.string,
  generatingError: PropTypes.string,
  generatingErrorMessages: PropTypes.arrayOf(
    PropTypes.shape({ message: PropTypes.string })
  ),
};

TemplateGenerator.defaultProps = {
  polling: false,
  dataUrl: null,
  generatingError: null,
  generatingErrorMessages: null,
};

export default TemplateGenerator;
