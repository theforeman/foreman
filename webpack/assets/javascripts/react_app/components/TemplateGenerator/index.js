import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { selectGeneratingProps } from './TemplateGeneratorSelectors';
import reducer from './TemplateGeneratorReducer';
import * as templateActions from './TemplateGeneratorActions';
import TemplateGenerator from './TemplateGenerator';

export const actions = templateActions;

// export reducers
export const reducers = { templates: reducer };

// map state to props
const mapStateToProps = state => selectGeneratingProps(state);

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export connected component
export default connect(mapStateToProps, mapDispatchToProps)(TemplateGenerator);
