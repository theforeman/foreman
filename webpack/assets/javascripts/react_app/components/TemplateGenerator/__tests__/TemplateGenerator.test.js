import { testComponentSnapshotsWithFixtures } from '../../../common/testHelpers';

import TemplateGenerator from '../TemplateGenerator';

const fixtures = {
  'render button if not polling and no errors': {
    data: {
      templateName: 'template',
    },
    polling: false,
    dataUrl: '/data/IDENTIFIER.json',
  },
  'render info alert if polling': {
    data: {
      templateName: 'template',
    },
    polling: true,
    dataUrl: '/data/IDENTIFIER.json',
  },
  'renders errors if there are some': {
    data: {
      templateName: 'template',
    },
    polling: false,
    generatingError: '422 unprocessable entity',
    generatingErrorMessages: [
      { message: 'Eh there was no method error during the render :(' },
    ],
  },
};

describe('TemplateGenerator', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(TemplateGenerator, fixtures));
});
