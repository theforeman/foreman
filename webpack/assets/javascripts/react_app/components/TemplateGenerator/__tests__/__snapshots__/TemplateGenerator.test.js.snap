// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateGenerator rendering render button if not polling and no errors 1`] = `
<Fragment>
  <Alert
    className=""
    onDismiss={null}
    title="Generating a report"
    type="success"
  >
    <AlertBody
      message="
  Generating of the report %s has been completed.
  Download should start automatically.
  In case it does not, please use the download button below.
"
    />
  </Alert>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="primary"
    disabled={false}
    href="/data/IDENTIFIER.json"
    target="_blank"
  >
    Download
  </Button>
</Fragment>
`;

exports[`TemplateGenerator rendering render info alert if polling 1`] = `
<Fragment>
  <Alert
    className=""
    onDismiss={null}
    title="Generating a report"
    type="info"
  >
    <AlertBody
      message="
  Report %s is now being generated, the download will start once it's done.
  You can come to this page later to get the results. The result is available for 24 hours.
"
    />
  </Alert>
</Fragment>
`;

exports[`TemplateGenerator rendering renders errors if there are some 1`] = `""`;
